lib/puppet/reports/datadog_reports.rb

```diff
@@ -110,6 +124,21 @@ def process
       event_data << "\n@@@\n"
     end
 
+    # Check for a running environment other than what is defined for check_environments in parameters
+    if CHECK_ENVIRONMENTS
+      if CHECK_ENVIRONMENTS.include?(@msg_environment) && @noop == false
+        environment_check_status = 0
+        environment_check_message = "#{@msg_host} is configured to use environment production"
+      else
+        environment_check_status = 2
+        environment_check_message = "#{@msg_host} is using environment #{@msg_environment} rather than one of:\n#{CHECK_ENVIRONMENTS.join(', ')}"
+      end
+    else
+      # Set to unknown status because we don't know if the environment is a valid one or not
+      environment_check_status = 3
```

Comment:
Personally my expectation is that "Don't change the result of my checks when I didn't tell you to check for this."

Reply:
I kind of agree. We get here if check_environments is not configured, so technically unknown makes sense. If I was me, except not the me who wants this feature, I'd want to be able to turn this feature off. However, if this is already idiomatic behavior for DD checks then being consistent overrides that.

Reply:
So this seems a natural state for this to be in to me.  This module will default to `['production']` as the check environment if you do not override the value here: https://github.com/krux/puppet-datadog-agent/blob/add_puppet_service_check/manifests/init.pp#L314

i.e. the only way to get to this state is if you setup this integration manually without using the puppet module or you manually modified the yaml to remove that or something.  Basically, if something is foo-bar and `check_environments` isn't configured even with the default value, then it goes to `unknown` check status on that system because it doesn't know what environment is supposed to be a valid one.

Also, it is important to think of checks and monitors as totally separate in Datadog. Just because an integration provides a check, or a custom check exists, it does not mean that it's connected to any monitor or alert.  When creating a monitor that uses the check in question, in this case `puppet.environment`, you have to say what scope it belongs in, i.e. `environment:prod`, `host:foo.bar.domain.tld`, `cluster_name:foo_prod`, except: `environment:dev`, etc.

So if you don't want the check to do anything, then don't make a monitor for it and you'll never hear from it.  This check is also working with already processed data so the overhead of having it there is very very small.  This happens only on the puppet server processing the report and it's already processing the report so "I want to turn this check off" doesn't really make sense.  If you don't want it or it's not applicable to you, then just don't make a monitor for it and there wouldn't be any negative effects.  That's consistent with checks from other integrations.  They're there but you don't worry about them unless you want to and you include them in a monitor.

Reply:
1. I do not have a super strong opinion about this. I have my own opinion, but I think the better decision maker for this question would be the actual maintainers of this integration.
2. I would prefer the default value of `production` to be built into this ruby script as well, but I don't know ruby enough to know whether that is possible and/or is a good idea. Having a default value in the puppet makes me feel much better about this anyways.
3. How does this handle the situation where I don't care about the puppet environment but want to be notified when puppet fails? I would have to set up monitoring to be notified when puppet fails, and put every possible environments in the list. Am I wrong on how that is being used? Doing a regex check may help that issue.

Overall, I don't think I am the best person to make this decision. I am content with this code as is and happy to let the maintainers reject the PR if they want to. The code is fine as is for Krux usage.

Reply:
The puppet failure and puppet environment checks are two separate checks, wholly independent of each other.  They show up in datadog as 2 separate and unrelated checks.  Consequently, if you want to monitor both checks because you care if puppet succeeded and you also care if the environment is configured correctly, 2 different monitors required because they're separate checks.  The custom check collects both service checks as it runs but both are fully independent of each other.

Reply:
I was thinking more of useless events spamming the event stream. If that's not an issue then I'm fine with it.

Reply:
As far as the unknown state goes, I think that's basically decent exception handling.  If we're in a "How did we get here?" state, the check goes into an `unknown` status rather than "everything's fine!" when it might not be.  Since you can't improve what you can't measure, I felt `unknown` was the right place for a state that you almost have to try to get yourself into and it would be apparent that something wasn't fully configured (and also tell you why in the check message) if you went looking for it.  The environments are determined by the YAML configuration, which is consistent with how everything check/integration at datadog works.  I felt it was appropriate for the YAML configuration to be where it's done at rather than hard coding anything in the ruby script, and puppet makes sure there's a default value there if you don't add one.